Sources/HaCTML/AttributeMap.swift

```diff
@@ -37,7 +37,26 @@ struct AttributeMap: ExpressibleByArrayLiteral {
     return attributeMap.values.map { attribute in attribute.key.apply(attribute.value) }
   }
 
+  var rendered: String? {
+    guard !isEmpty else {
+      return nil
+    }
+
+    return renderedValues.flatMap(AttributeMap.renderAppliedAttribute).joined(separator: " ")
+  }
+
   var isEmpty: Bool {
     return self.attributeMap.isEmpty
   }
+
+  static func renderAppliedAttribute(_ attribute: AppliedAttribute) -> String? {
```

Comment:
I feel like adding these to the struct here clutter it, but there isn't much else to do and if it makes the codebase easier to parse then I'm for it.

Reply:
Agreed this is not an ideal place for this but a better one didn't come to mind. I am keen to avoid top-level functions  in packages where possible, I think mostly from a hygiene point of view.
Overall, now that we're on Swift 4, perhaps we could take another look at the layout of the Attribute stuff and tidy a few things up, I still find it tricky to navigate all these different Attribute bits and pieces